common/transport/amqp/src/receiver_link.ts

```diff
@@ -80,98 +149,236 @@ export class ReceiverLink  extends EventEmitter implements AmqpLink {
+            //
+            // According to the rhea maintainers, one can depend on that fact that no actual network activity
+            // will occur until the nextTick() after the call to open_receiver.  Because of that, one can
+            // put the event handlers on the rhea link returned from the open_receiver call and be assured
+            // that the listeners are in place BEFORE any possible events will be emitted on the link.
+            //
+            this._rheaReceiver = this._rheaSession.open_receiver(this._combinedOptions);
+            manageReceiverHandlers('on');
+          },
+          receiverOpenEvent: (context: EventContext) => {
+            debug('In receiver attaching state - open event for ' + context.receiver.name);
+            let callback = this._attachingCallback;
+            this._attachingCallback = null;
+            this._fsm.transition('attached', callback);
+          },
+          receiverErrorEvent: (context: EventContext) => {
+            debug('In receiver attaching state - error event for ' + context.receiver.name + ' error is: ' + this._getErrorName(context.receiver.error));
+            this._indicatedError = context.receiver.error;
+            //
+            // We don't transition at this point in that we are guaranteed that the error will be followed by a receiver_close
+            // event.
+            //
+          },
+          receiverCloseEvent: (context: EventContext) => {
+            debug('in receiver attaching state - close event for ' + context.receiver.name);
+            //
+            // We enabled the event listeners on the onEnter handler.  They are to stay alive until we
+            // are about to transition to the detached state.
+            // We are about to transition to the detached state, so clean up.
+            //
+            manageReceiverHandlers('removeListener');
+            let error = this._indicatedError;
+            let callback = this._attachingCallback;
+            this._indicatedError = undefined;
+            this._attachingCallback = undefined;
+            this._receiverCloseOccurred = true;
+            this._fsm.transition('detached', callback, error);
+          },
+          attach: (null),
+          detach: (callback, err) => {
+            /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_06_002: [If the `detach` method is invoked on the `ReceiverLink` while still attaching, the ReceiverLink shall detach.  It will indicate the error to the callback for the `detach` as well as the callback to the `attach`.] */
+            debug('Detaching while attaching of rhea receiver link ' + this._linkAddress);
+            manageReceiverHandlers('removeListener');
+            //
+            // We may have a callback outstanding on the request that started the attaching.
+            // We will signal to that callback that an error occurred. We will also invoke the callback supplied
+            // for this detach.
```

Comment:
wait... we have 2 callbacks to call here? that's really weird. there should be a way to call a single callback that unrolls all the way to the original operation.

Reply:
See reponse to 2 below.

---
In reply to: [209359414](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209359414) [](ancestors = 209359414)